### DISCUSSION TITLE
Moo - A Twitter clone (mainly for shitposting)

### DISCUSSION THREAD

**Discussion by championswimmer** [scaleracademy/open-source-projects#81] [Projects]
## Projects 

- Backend (in NodeJS) https://github.com/scaleracademy/twitter-backend-node
- Backend (in Java) https://github.com/scaleracademy/twitter-backend-java

## Summary 

A twitter-like social media site where anyone can post 240-character texts (future: images), and people can reply, like or reshare the posts. 
Users can follow each other. The home feed shows posts from users that they follow. 
Users will be notified when someone likes their posts, replies or reshares them and when someone follows them.
Users and posts can be searched.

### Future Scope
- Report posts ✅
- User's can send personal messages to each other ❓
- Allow accounts to be private (only visible to followers) ❓
- Turn on notifications for a particular user ❓
- Bookmarking posts ❓
- Topics ❓

### 1. Users 
- username 
- avatar (display pic) 
- name 
- email id 
- contact no 
- bio 

### 2. Posts 
- author 
- text 
- image (upto 4) 
- likes 
- reposts 
- comments 
- inReplyTo -> another post (or null) 
- timestamp 
- mentions (via @user ) 
- hashtags (via #tags ) 

### 3. Notification 
- to 
- post (if it is a like/reply/share) 
- message 

### 4. Hashtags 
- title (lowercase strings)

**Comment by championswimmer** [comment]
# Database Design 

### Tables 

#### 1. `users` 
| column | definition
|--------------|------------------------------------|
| id              | uuid                           |
| username        | string(30) - only alphanumeric |
| name            | string (50)                    |
| avatar          | string (url)                   |
| bio             | text (240)                     |
| follower_count  | bigint                         |
| following_count | bigint                         |
| verified        | boolean                        |

#### 2. `posts` 
| column | definition
|--------------|------------------------------------|
| id           | uuid                               |
| text         | string (240) chars                 |
| author_id    | uuid - foreignkey(user.id)         |
| images       | array<string(url)> - max 4         |
| like_count   | bigint                             |
| repost_count | bigint                             |
| orig_post_id | uuid (self reference, can be null) |
| reply_to_id  | uuid (self reference, can be null) |
| hashtags     | array<string>                      |
| mentions     | array<{name: string, id: uuid}>    |


#### 3. `hashtags` 

| column            | definition                  |
|-------------------|-----------------------------|
| id                | uuid                        |
| tag               | text (only a-z,0-9,_ chars) |
| recent_post_count | bigint                      |

#### 4. `likes` 
| column  | definition                   |
|---------|------------------------------|
| id      | uuid                         |
| post_id | uuid - foreign key (post.id) |
| user_id | uuid - foreign key (user.id) |

#### 5. `hashtag_posts` 

| column     | definition                      |
|------------|---------------------------------|
| id         | uuid                            |
| hashtag_id | uuid - foreign key (hashtag.id) |
| post_id    | uuid - foreign key (post.id)    |

#### 6. `user_followings` 
| column      | definition                   |
|-------------|------------------------------|
| id          | uuid                         |
| follower_id | uuid - foreign key (user.id) |
| followee_id | uuid - foreign key (user.id) |

#### 7. `post_mentions` 
TBD

**Reply by ohbus** [reply]
Users table should have a password field
And posts should have a timestamp

**Reply by ohbus** [reply]
![Schema Draft](https://user-images.githubusercontent.com/13291222/121964106-a573ad00-cd88-11eb-8510-dfac87552568.png)

Have Completed the schema, can you confirm @championswimmer?

**Reply by championswimmer** [reply]
We will have `createdAt` and `updatedAt` on all tables actually, so didn't explicitly mentioned it 

Looks good otherwise @ohbus

**Reply by ohbus** [reply]
I Will update the schema accordingly then, before we hit the show tonight, _hopefully_.

**Reply by PCS0725** [reply]
@ohbus  Can you please mention the tool you used to create the data schema diagram?

**Reply by iamsriranga** [reply]
@ohbus u can use this [Lucid Chart](https://www.lucidchart.com/pages/)

**Reply by ohbus** [reply]
> @ohbus Can you please mention the tool you used to create the data schema diagram?

This is [MySQL Workbench](https://dev.mysql.com/downloads/workbench/) ER Modelling

**Reply by ohbus** [reply]
> @ohbus u can use this [Lucid Chart](https://www.lucidchart.com/pages/)

Thanks, @saideepesh000 for the awesome suggestion! I'll surely check it out.

**Reply by mkumar9009** [reply]
will recent_post_count be reset after certain time .. if no new post is coming for that hashtag.. ? if it will contain the overall count the name shall be clear then

**Reply by ohbus** [reply]
We are certainly not clear about what interval we should be taking here for this implementation as of now.

Maybe @championswimmer can suggest us this, we already had a discussion but didn't come to any conclusion.

**Reply by ViruSutar** [reply]
hey can you tell me why you have used like_count column in the post table .we can count that likes from likes table can't we ?

**Comment by championswimmer** [comment]
# API Design 

## `auth` 

### ✅ `POST /auth/login` 
Login and get auth token 

## `users` 

### ✅ `GET /users/@{username}`  
Get details of given user by username

### ✅ `GET /users/{userid}` 
Get details of given user by userid

### ✅ `POST /users` 
Create a new user 

### ✅ `PATCH /users/{userid}` 🔒
Update bio/name/image etc of an user 

### `PUT /users/{userid}/follow` 🔒 
Follow the given user 

### `DELETE /users/{userid}/follow` 🔒 
Un-follow the given user 

### `GET /users/{userid}/followers` 📃
Get a list of all followers of this user

### `GET /users/{userid}/followees` 📃
Get a list of all followees of this user

## `posts` 

### `GET /posts`  📃

`query_params` 
| query    | definition            |
|----------|-----------------------|
| by       | username              |
| reply_to | post uuid             |
| hashtag | tag             |

### `GET /posts/{postid}`
Get details of a post 

```javascript 
{
	"text": "....",
	"like_count": 999,
	"replies": [ // first 50
		{
			"text": "....",
			"like_count": 999,
		},
		{
			"text": "....",
			"like_count": 999,
		}
		
	]
}

```

### `POST /posts` 🔒 
Create a new post  

### `DELETE /posts/{postid}` 🔒
Delete a given post id

### `PUT /posts/{postid}/like` 🔒
Like the given post

### `DELETE /posts/{postid}/like` 🔒
Un-like the given post

## `hashtags` 

### `GET /hashtags`  📃 
Top hashtags (default top 10) 

### `GET /hashtags/{tag}/posts`  📃 
All posts of this given hashtag

**Comment by championswimmer** [comment]
Anyone seeing this discussion, there is vote on our Discord server about which framework we will use for the backend. 
Please join and vote, if you'd like to 

https://canary.discord.com/channels/780066247601291285/787377926722945035/854365075036045322 

<img width="710" alt="image" src="https://user-images.githubusercontent.com/1327050/122070056-48710900-ce13-11eb-9785-e529995158fc.png">

**Reply by amsanket22** [reply]
I would suggest Node, NestJS to be specific, Typescript is good

**Reply by championswimmer** [reply]
@amsanket22 yes I plan to do with NestJS only :D

**Reply by sarthak090** [reply]
What about firebase

**Reply by Adrenalinerush07** [reply]
I would suggest Nest.js, GraphQl, and MongoDB

**Reply by championswimmer** [reply]
Since we are using our own backend, why do we need Firebase ?

**Comment by VenomFate-619** [comment]
> ## Projects
> * Backend (in NodeJS) https://github.com/scaleracademy/twitter-backend-node
> * Backend (in Java) https://github.com/scaleracademy/twitter-backend-java
> 
> ## Summary
> A twitter-like social media site where anyone can post 240-character texts (future: images), and people can reply, like or reshare the posts.
> Users can follow each other. The home feed shows posts from users that they follow.
> Users will be notified when someone likes their posts, replies or reshares them and when someone follows them.
> Users and posts can be searched.
> 
> ### Future Scope
> * Report posts white_check_mark
> * User's can send personal messages to each other question
> * Allow accounts to be private (only visible to followers) question
> * Turn on notifications for a particular user question
> * Bookmarking posts question
> * Topics question
> 
> ## Entities (in product terms)
> ### 1. Users
> * username
> * avatar (display pic)
> * name
> * email id
> * contact no
> * bio
> 
> ### 2. Posts
> * author
> * text
> * image (upto 4)
> * likes
> * reposts
> * comments
> * inReplyTo -> another post (or null)
> * timestamp
> * mentions (via @user )
> * hashtags (via #tags )
> 
> ### 3. Notification
> * to
> * post (if it is a like/reply/share)
> * message
> 
> ### 4. Hashtags
> * title (lowercase strings)

how will manage to mention users in a post means how we can give suggestion

**Reply by bhargavgohil25** [reply]
Whenever, someone tries to write "@bh" we will pass information of all the users that starts from "bh" and including the user who is assigned to that post or comment or anything.
So, I think there will be an API call for every new character that enter.
This is my upfront idea. 
If this isn't how mentioning works please correct me.

**Comment by kushal-khare-official** [comment]
Hey guys, I am looking for some good projects to contribute to this October. Was wondering if this project is participating in Hactoberfest or not.
Regardless of that, I wish to contribute. Also I wanted to know if I can work on an endpoint from the above discussion for which an issue hasn't been raised yet.

**Reply by ohbus** [reply]
You may raise an issue in any of the repositories that you may find comfortable for yourself to contribute.

Respective maintainers will help you guide along the way.

**Comment by ViruSutar** [comment]
hii , can i still contribute to this project ?

**Reply by championswimmer** [reply]
Sure you can
